terraform/eval_for_each.go

```diff
@@ -58,16 +58,20 @@ func evaluateForEachExpressionValue(expr hcl.Expression, ctx EvalContext, allowU
 
 	forEachVal, forEachDiags := expr.Value(hclCtx)
 	diags = diags.Append(forEachDiags)
-	if forEachVal.ContainsMarked() {
+
+	// If a whole map is marked, or a set contains marked values (which means the set is then marked)
+	// give an error diagnostic as this value cannot be used in for_each
+	if forEachVal.IsMarked() {
 		diags = diags.Append(&hcl.Diagnostic{
 			Severity:    hcl.DiagError,
 			Summary:     "Invalid for_each argument",
-			Detail:      "Sensitive variables, or values derived from sensitive variables, cannot be used as for_each arguments. If used, the sensitive value could be exposed as a resource instance key.",
```

Comment:
Could this "or derived by" be left to communicate the function-result situation? Unsure.

Reply:
I like that phrase, personally, and think it covers the function-result situation, but not in a "blocking merge" way!

Reply:
Should we use a term like "sensitive values" here? I think this diagnostic applies to all of sensitive variables, outputs, and attributes (with the experiment enabled).

Reply:
@alisdair excellent point, "values" is correct rather than variables! And thanks for the input @mildwonkey, I'll add it back